test/utils/auth.test.js

```diff
@@ -0,0 +1,92 @@
+const nock = require('nock');
+const axios = require('axios');
+const faker = require('faker');
+
+const config = require('../../lib/config');
+const { setV5Token } = require('../../lib/utils/auth');
+
+const url = faker.internet.url();
+const clientId = faker.random.uuid();
+const clientSecret = faker.internet.password();
+
+// test helpers
+
+/**
+ * Since OAuth Bearer tokens don't have a defined structure, we will
+ * generate alphanumeric strings of length 24.
+ *
+ * https://www.oauth.com/oauth2-servers/access-tokens/access-token-response/
+ *
+ * @returns {string} testAccessToken
+ *
+ * @example
+ * getTestAccessToken() // => mjpyxya9signd3g9zu1s1zkb
+ */
+const getTestAccessToken = () => faker.random.alphaNumeric(24);
+const mockV5AuthAPI = ({ testToken = getTestAccessToken(), createdAt = Date.now() }) =>
+  nock(url)
+    .post('/api/v5/auth/token', {
+      grant_type: 'client_credentials',
+    })
+    .basicAuth({ user: clientId, pass: clientSecret })
+    .reply(200, {
+      access_token: testToken,
+      token_type: 'Bearer',
+      expires_in: 7200,
+      created_at: createdAt,
+    });
+
+describe('setV5Token', () => {
+  beforeEach(() => {
+    config.set('piiServerUrl', url);
+    config.set('clientId', clientId);
+    config.set('clientSecret', clientSecret);
+    axios.defaults.headers.common.Authorization = undefined;
+    axios.defaults.tokenData = undefined;
+  });
+  // clean up after all tests in this file have run.
+  afterAll(() => {
+    axios.defaults.headers.common.Authorization = undefined;
+    axios.defaults.tokenData = undefined;
+  });
+
+  it('should set the accessToken', async () => {
+    const expectedTestToken = getTestAccessToken();
+    mockV5AuthAPI({ testToken: expectedTestToken });
+
+    await setV5Token();
+
+    expect(axios.defaults.headers.common.Authorization).toEqual(`Bearer ${expectedTestToken}`);
+  });
+
+  it('should cache the token for subsequent calls', async () => {
+    const expectedTestToken = getTestAccessToken();
+
+    mockV5AuthAPI({ testToken: expectedTestToken });
+
+    await setV5Token();
+    await setV5Token();
+
+    expect(axios.defaults.headers.common.Authorization).toEqual(`Bearer ${expectedTestToken}`);
```

Comment:
Question: Kind of the same concern as before: Given we're caching the most recent bearer token, do we even need the race condition mitigation strategy? My gut says "Yes" but I haven't worked it out yet…

Reply:
I think your concern is valid, it would be inefficient to allow multiple auth calls to be inflight concurrently. I will refactor accordingly.